app/common/state/contextMenuState.js

```diff
@@ -0,0 +1,159 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const assert = require('assert')
+const eventUtil = require('../../../js/lib/eventUtil.js')
+const appActions = require('../../../js/actions/appActions.js')
+const windowAction = require('../../../js/actions/windowActions.js')
+const config = require('../../../js/constants/config.js')
+const CommonMenu = require('../commonMenu.js')
+const locale = require('../../../js/l10n.js')
+const { makeImmutable, isMap } = require('./immutableUtil')
+
+const validateAction = function (action) {
+  action = makeImmutable(action)
+  assert.ok(isMap(action), 'action must be an Immutable.Map')
+  return action
+}
+
+const validateState = function (state) {
+  state = makeImmutable(state)
+  assert.ok(isMap(state), 'state must be an Immutable.Map')
+  return state
+}
+
+const contextMenuState = {
+  setContextMenu: (state, action) => {
+    action = validateAction(action)
+    state = validateState(state)
+
+    if (!action.get('detail')) {
+      if (state.getIn(['contextMenuDetail', 'type']) === 'hamburgerMenu') {
+        state = state.set('hamburgerMenuWasOpen', true)
+      } else {
+        state = state.set('hamburgerMenuWasOpen', false)
+      }
+      state = state.delete('contextMenuDetail')
+    } else {
+      if (!(action.getIn(['detail', 'type']) === 'hamburgerMenu' && state.get('hamburgerMenuWasOpen'))) {
+        state = state.set('contextMenuDetail', action.get('detail'))
+      }
+      state = state.set('hamburgerMenuWasOpen', false)
+    }
+
+    return state
+  },
+
+  onLongBacHistory: (state, action) => {
```

Comment:
I think these two methods probably make more sense in the reducer. The state helpers are mostly just getters/setters for the state. I'm not adamantly opposed to leaving them here, but if they stay the params should be tabId, partitionNumber and history instead of action